File: backend/feature/blog/service/BlogService.cs
using System.Text.RegularExpressions;
using Data;
using Entities;
using Microsoft.EntityFrameworkCore;
using Models;

namespace Services
{
    public record BlogPostSummaryResponse(
        int Id,
        string Title,
        string Slug,
        string Excerpt,
        string Category,
        string CoverImageUrl,
        string AuthorName,
        DateTime PublishedAt,
        int ReadTimeMinutes,
        bool IsFeatured
    );

    public record BlogPostResponse(
        int Id,
        string Title,
        string Slug,
        string Excerpt,
        string Category,
        string CoverImageUrl,
        string? CoverImageObjectName,
        string AuthorName,
        int ReadTimeMinutes,
        bool IsFeatured,
        bool IsPublished,
        DateTime? PublishedAt,
        List<string> Tags,
        List<string> Highlights,
        List<string> Paragraphs,
        DateTime CreatedAt,
        DateTime UpdatedAt
    );

    public record BlogPostDetailResponse(
        int Id,
        string Title,
        string Slug,
        string Excerpt,
        string Category,
        string CoverImageUrl,
        string AuthorName,
        DateTime PublishedAt,
        int ReadTimeMinutes,
        List<string> Tags,
        List<string> Highlights,
        List<string> Paragraphs,
        List<BlogPostSummaryResponse> RelatedPosts
    );

    public class CreateBlogPostRequest
    {
        public string Title { get; set; } = string.Empty;
        public string Excerpt { get; set; } = string.Empty;
        public string Category { get; set; } = string.Empty;
        public string CoverImageUrl { get; set; } = string.Empty;
        public string? CoverImageObjectName { get; set; }
        public string AuthorName { get; set; } = string.Empty;
        public int ReadTimeMinutes { get; set; } = 5;
        public bool IsFeatured { get; set; }
        public bool IsPublished { get; set; } = true;
        public DateTime? PublishedAt { get; set; }
        public List<string> Tags { get; set; } = [];
        public List<string> Highlights { get; set; } = [];
        public List<string> Paragraphs { get; set; } = [];
    }

    public class UpdateBlogPostRequest : CreateBlogPostRequest
    {
        public int Id { get; set; }
    }

    public class BlogService
    {
        private readonly AppDbContext _db;

        public BlogService(AppDbContext db)
        {
            _db = db;
        }

        public async Task<BlogPostResponse> CreateBlogPostAsync(CreateBlogPostRequest request)
        {
            var post = new BlogPost();
            await ApplyRequestAsync(post, request);
            post.CreatedAt = DateTime.UtcNow;
            post.UpdatedAt = post.CreatedAt;

            await _db.BlogPosts.AddAsync(post);
            await _db.SaveChangesAsync();

            return MapAdmin(post);
        }

        public async Task<BlogPostResponse?> UpdateBlogPostAsync(UpdateBlogPostRequest request)
        {
            var post = await _db.BlogPosts.FirstOrDefaultAsync(item => item.Id == request.Id);

            if (post is null)
            {
                return null;
            }

            await ApplyRequestAsync(post, request);
            post.UpdatedAt = DateTime.UtcNow;

            await _db.SaveChangesAsync();

            return MapAdmin(post);
        }

        public async Task DeleteBlogPostAsync(int id)
        {
            var post = await _db.BlogPosts.FirstOrDefaultAsync(item => item.Id == id);

            if (post is null)
            {
                return;
            }

            _db.BlogPosts.Remove(post);
            await _db.SaveChangesAsync();
        }

        public async Task<PaginatedResult<BlogPostResponse>> GetAdminBlogPostsAsync(
            int page = 1,
            int pageSize = 12,
            string? search = null,
            bool? isPublished = null)
        {
            var normalizedPage = Math.Max(1, page);
            var normalizedPageSize = Math.Max(1, pageSize);

            var query = _db.BlogPosts.AsNoTracking().AsQueryable();

            if (!string.IsNullOrWhiteSpace(search))
            {
                var normalizedSearch = search.Trim().ToLowerInvariant();
                query = query.Where(item =>
                    item.Title.ToLower().Contains(normalizedSearch) ||
                    item.Excerpt.ToLower().Contains(normalizedSearch) ||
                    item.Category.ToLower().Contains(normalizedSearch) ||
                    item.AuthorName.ToLower().Contains(normalizedSearch));
            }

            if (isPublished.HasValue)
            {
                query = query.Where(item => item.IsPublished == isPublished.Value);
            }

            var totalCount = await query.CountAsync();
            var items = await query
                .OrderByDescending(item => item.UpdatedAt)
                .Skip((normalizedPage - 1) * normalizedPageSize)
                .Take(normalizedPageSize)
                .ToListAsync();

            return new PaginatedResult<BlogPostResponse>
            {
                Items = items.Select(MapAdmin).ToList(),
                TotalCount = totalCount,
                Page = normalizedPage,
                PageSize = normalizedPageSize,
                TotalPages = (int)Math.Ceiling(totalCount / (double)normalizedPageSize)
            };
        }

        public async Task<PaginatedResult<BlogPostSummaryResponse>> GetBlogPostsAsync(
            int page = 1,
            int pageSize = 9,
            string? search = null,
            string? category = null,
            bool featuredOnly = false)
        {
            var normalizedPage = Math.Max(1, page);
            var normalizedPageSize = Math.Max(1, pageSize);

            var query = _db.BlogPosts
                .AsNoTracking()
                .Where(item => item.IsPublished)
                .AsQueryable();

            if (!string.IsNullOrWhiteSpace(search))
            {
                var normalizedSearch = search.Trim().ToLowerInvariant();
                query = query.Where(item =>
                    item.Title.ToLower().Contains(normalizedSearch) ||
                    item.Excerpt.ToLower().Contains(normalizedSearch) ||
                    item.Category.ToLower().Contains(normalizedSearch) ||
                    item.AuthorName.ToLower().Contains(normalizedSearch));
            }

            if (!string.IsNullOrWhiteSpace(category))
            {
                var normalizedCategory = category.Trim().ToLowerInvariant();
                query = query.Where(item => item.Category.ToLower() == normalizedCategory);
            }

            if (featuredOnly)
            {
                query = query.Where(item => item.IsFeatured);
            }

            var totalCount = await query.CountAsync();
            var items = await query
                .OrderByDescending(item => item.IsFeatured)
                .ThenByDescending(item => item.PublishedAt ?? item.UpdatedAt)
                .Skip((normalizedPage - 1) * normalizedPageSize)
                .Take(normalizedPageSize)
                .ToListAsync();

            return new PaginatedResult<BlogPostSummaryResponse>
            {
                Items = items.Select(MapSummary).ToList(),
                TotalCount = totalCount,
                Page = normalizedPage,
                PageSize = normalizedPageSize,
                TotalPages = (int)Math.Ceiling(totalCount / (double)normalizedPageSize)
            };
        }

        public async Task<BlogPostDetailResponse?> GetBlogPostBySlugAsync(string slug)
        {
            var normalizedSlug = (slug ?? string.Empty).Trim().ToLowerInvariant();

            if (normalizedSlug.Length == 0)
            {
                return null;
            }

            var post = await _db.BlogPosts
                .AsNoTracking()
                .FirstOrDefaultAsync(item => item.IsPublished && item.Slug.ToLower() == normalizedSlug);

            if (post is null)
            {
                return null;
            }

            var relatedPosts = await _db.BlogPosts
                .AsNoTracking()
                .Where(item =>
                    item.IsPublished &&
                    item.Id != post.Id &&
                    item.Category.ToLower() == post.Category.ToLower())
                .OrderByDescending(item => item.IsFeatured)
                .ThenByDescending(item => item.PublishedAt ?? item.UpdatedAt)
                .Take(3)
                .ToListAsync();

            return new BlogPostDetailResponse(
                post.Id,
                post.Title,
                post.Slug,
                post.Excerpt,
                post.Category,
                post.CoverImageUrl,
                post.AuthorName,
                post.PublishedAt ?? post.UpdatedAt,
                post.ReadTimeMinutes,
                post.Tags.ToList(),
                post.Highlights.ToList(),
                post.Paragraphs.ToList(),
                relatedPosts.Select(MapSummary).ToList()
            );
        }

        private async Task ApplyRequestAsync(BlogPost post, CreateBlogPostRequest request)
        {
            var title = (request.Title ?? string.Empty).Trim();
            var excerpt = (request.Excerpt ?? string.Empty).Trim();
            var category = (request.Category ?? string.Empty).Trim();
            var coverImageUrl = (request.CoverImageUrl ?? string.Empty).Trim();
            var authorName = (request.AuthorName ?? string.Empty).Trim();
            var paragraphs = NormalizeOrderedTextList(request.Paragraphs);

            if (title.Length == 0)
            {
                throw new ArgumentException("Title is required.");
            }

            if (excerpt.Length == 0)
            {
                throw new ArgumentException("Excerpt is required.");
            }

            if (category.Length == 0)
            {
                throw new ArgumentException("Category is required.");
            }

            if (coverImageUrl.Length == 0)
            {
                throw new ArgumentException("Cover image is required.");
            }

            if (authorName.Length == 0)
            {
                throw new ArgumentException("Author name is required.");
            }

            if (paragraphs.Count == 0)
            {
                throw new ArgumentException("Add at least one article paragraph.");
            }

            post.Title = title;
            post.Slug = await EnsureUniqueSlugAsync(title, post.Id == 0 ? null : post.Id);
            post.Excerpt = excerpt;
            post.Category = category;
            post.CoverImageUrl = coverImageUrl;
            post.CoverImageObjectName = NullIfEmpty(request.CoverImageObjectName);
            post.AuthorName = authorName;
            post.ReadTimeMinutes = Math.Max(1, request.ReadTimeMinutes);
            post.IsFeatured = request.IsFeatured;
            post.IsPublished = request.IsPublished;
            post.PublishedAt = request.IsPublished
                ? request.PublishedAt?.ToUniversalTime() ?? post.PublishedAt ?? DateTime.UtcNow
                : request.PublishedAt?.ToUniversalTime();
            post.Tags = NormalizeDistinctTextList(request.Tags);
            post.Highlights = NormalizeOrderedTextList(request.Highlights);
            post.Paragraphs = paragraphs;
        }

        private async Task<string> EnsureUniqueSlugAsync(string title, int? excludeId)
        {
            var baseSlug = GenerateSlug(title);
            var slug = baseSlug;
            var suffix = 2;

            while (await _db.BlogPosts.AnyAsync(item => item.Slug == slug && (!excludeId.HasValue || item.Id != excludeId.Value)))
            {
                slug = $"{baseSlug}-{suffix}";
                suffix++;
            }

            return slug;
        }

        private static string GenerateSlug(string title)
        {
            var slug = (title ?? string.Empty).Trim().ToLowerInvariant();
            slug = Regex.Replace(slug, @"[^a-z0-9\s-]", string.Empty);
            slug = Regex.Replace(slug, @"\s+", "-");
            slug = Regex.Replace(slug, @"-+", "-");
            slug = slug.Trim('-');

            return slug.Length > 0 ? slug : "blog-post";
        }

        private static List<string> NormalizeDistinctTextList(IEnumerable<string>? items)
        {
            return (items ?? [])
                .Where(item => !string.IsNullOrWhiteSpace(item))
                .Select(item => item.Trim())
                .Distinct(StringComparer.OrdinalIgnoreCase)
                .ToList();
        }

        private static List<string> NormalizeOrderedTextList(IEnumerable<string>? items)
        {
            return (items ?? [])
                .Where(item => !string.IsNullOrWhiteSpace(item))
                .Select(item => item.Trim())
                .ToList();
        }

        private static string? NullIfEmpty(string? value)
        {
            var normalized = (value ?? string.Empty).Trim();
            return normalized.Length == 0 ? null : normalized;
        }

        private static BlogPostResponse MapAdmin(BlogPost post)
        {
            return new BlogPostResponse(
                post.Id,
                post.Title,
                post.Slug,
                post.Excerpt,
                post.Category,
                post.CoverImageUrl,
                post.CoverImageObjectName,
                post.AuthorName,
                post.ReadTimeMinutes,
                post.IsFeatured,
                post.IsPublished,
                post.PublishedAt,
                post.Tags.ToList(),
                post.Highlights.ToList(),
                post.Paragraphs.ToList(),
                post.CreatedAt,
                post.UpdatedAt
            );
        }

        private static BlogPostSummaryResponse MapSummary(BlogPost post)
        {
            return new BlogPostSummaryResponse(
                post.Id,
                post.Title,
                post.Slug,
                post.Excerpt,
                post.Category,
                post.CoverImageUrl,
                post.AuthorName,
                post.PublishedAt ?? post.UpdatedAt,
                post.ReadTimeMinutes,
                post.IsFeatured
            );
        }
    }
}

File: backend-nestjs/src/blog/blog.service.ts
import { BadRequestException, Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { BlogPost } from './entities/blog-post.entity';
import { paginated, toInt } from '../common/api-contract';

@Injectable()
export class BlogService {
  constructor(
    @InjectRepository(BlogPost)
    private blogRepository: Repository<BlogPost>,
  ) {}

  async findAllAdmin(page = 1, pageSize = 12, search?: string, isPublished?: boolean) {
    page = toInt(page, 1);
    pageSize = toInt(pageSize, 12);
    const qb = this.blogRepository.createQueryBuilder('blog');
    if (search) {
      qb.andWhere('(blog.title ILIKE :search OR blog.excerpt ILIKE :search OR blog.category ILIKE :search OR blog.author_name ILIKE :search)', { search: `%${search}%` });
    }
    if (typeof isPublished === 'boolean') qb.andWhere('blog.is_published = :isPublished', { isPublished });
    const [rows, total] = await qb.orderBy('blog.updatedAt', 'DESC').skip((page - 1) * pageSize).take(pageSize).getManyAndCount();
    return paginated(rows.map((post) => this.mapAdmin(post)), total, page, pageSize);
  }

  async findAllPublic(page = 1, pageSize = 9, search?: string, category?: string, featuredOnly = false) {
    page = toInt(page, 1);
    pageSize = toInt(pageSize, 9);
    const qb = this.blogRepository.createQueryBuilder('blog').where('blog.is_published = true');
    if (search) {
      qb.andWhere('(blog.title ILIKE :search OR blog.excerpt ILIKE :search OR blog.category ILIKE :search OR blog.author_name ILIKE :search)', { search: `%${search}%` });
    }
    if (category) qb.andWhere('lower(blog.category) = :category', { category: category.toLowerCase() });
    if (featuredOnly) qb.andWhere('blog.is_featured = true');
    const [rows, total] = await qb
      .orderBy('blog.isFeatured', 'DESC')
      .addOrderBy('blog.publishedAt', 'DESC', 'NULLS LAST')
      .addOrderBy('blog.updatedAt', 'DESC')
      .skip((page - 1) * pageSize)
      .take(pageSize)
      .getManyAndCount();
    return paginated(rows.map((post) => this.mapSummary(post)), total, page, pageSize);
  }

  async findOne(idOrSlug: number | string) {
    const where = Number.isFinite(Number(idOrSlug))
      ? { id: Number(idOrSlug) }
      : { slug: String(idOrSlug) };
    const post = await this.blogRepository.findOneBy(where as any);
    if (!post) return null;
    const related = await this.blogRepository.find({
      where: { isPublished: true, category: post.category },
      order: { isFeatured: 'DESC', publishedAt: 'DESC' },
      take: 3,
    });
    return {
      id: post.id,
      title: post.title,
      slug: post.slug,
      excerpt: post.excerpt,
      category: post.category,
      coverImageUrl: post.coverImageUrl,
      authorName: post.authorName,
      publishedAt: post.publishedAt ?? post.updatedAt,
      readTimeMinutes: post.readTimeMinutes,
      tags: post.tags ?? [],
      highlights: post.highlights ?? [],
      paragraphs: post.paragraphs ?? [],
      relatedPosts: related.filter((item) => item.id !== post.id).slice(0, 3).map((item) => this.mapSummary(item)),
    };
  }

  async findPublicBySlug(slug: string) {
    const post = await this.blogRepository.createQueryBuilder('blog')
      .where('LOWER(blog.slug) = :slug', { slug: slug.trim().toLowerCase() })
      .andWhere('blog.is_published = true')
      .getOne();
    if (!post) throw new NotFoundException('Blog post not found');
    const related = await this.blogRepository.find({ where: { isPublished: true, category: post.category }, order: { isFeatured: 'DESC', publishedAt: 'DESC' }, take: 4 });
    return { id: post.id, title: post.title, slug: post.slug, excerpt: post.excerpt, category: post.category, coverImageUrl: post.coverImageUrl, authorName: post.authorName, publishedAt: post.publishedAt ?? post.updatedAt, readTimeMinutes: post.readTimeMinutes, tags: post.tags ?? [], highlights: post.highlights ?? [], paragraphs: post.paragraphs ?? [], relatedPosts: related.filter((item) => item.id !== post.id).slice(0, 3).map((item) => this.mapSummary(item)) };
  }

  async create(dto: any) {
    const post = this.blogRepository.create(dto as object);
    await this.applyDefaults(post);
    return this.mapAdmin(await this.blogRepository.save(post));
  }

  async update(id: number, dto: any) {
    const post = await this.blogRepository.findOne({ where: { id } });
    if (!post) throw new NotFoundException('Post not found');
    Object.assign(post, dto);
    await this.applyDefaults(post);
    return this.mapAdmin(await this.blogRepository.save(post));
  }

  async delete(id: number) {
    const post = await this.blogRepository.findOne({ where: { id } });
    if (post) await this.blogRepository.delete(id);
  }

  private async applyDefaults(post: BlogPost) {
    post.title = `${post.title ?? ''}`.trim();
    post.excerpt = `${post.excerpt ?? ''}`.trim();
    post.category = `${post.category ?? ''}`.trim();
    post.coverImageUrl = `${post.coverImageUrl ?? ''}`.trim();
    post.authorName = `${post.authorName ?? ''}`.trim();
    post.paragraphs = (post.paragraphs ?? []).map((item) => item?.trim()).filter(Boolean);
    if (!post.title) throw new BadRequestException('Title is required.');
    if (!post.excerpt) throw new BadRequestException('Excerpt is required.');
    if (!post.category) throw new BadRequestException('Category is required.');
    if (!post.coverImageUrl) throw new BadRequestException('Cover image is required.');
    if (!post.authorName) throw new BadRequestException('Author name is required.');
    if (!post.paragraphs.length) throw new BadRequestException('Add at least one article paragraph.');
    const base = this.slugify(post.title);
    let slug = base;
    let suffix = 2;
    while (await this.blogRepository.createQueryBuilder('blog').where('blog.slug = :slug', { slug }).andWhere(post.id ? 'blog.id != :id' : '1=1', { id: post.id }).getExists()) slug = `${base}-${suffix++}`;
    post.slug = slug;
    post.readTimeMinutes = Math.max(1, Number(post.readTimeMinutes) || 5);
    post.isPublished = post.isPublished ?? true;
    post.publishedAt = post.isPublished ? (post.publishedAt ?? new Date()) : post.publishedAt;
    post.tags = [...new Set((post.tags ?? []).map((item) => item?.trim()).filter(Boolean))];
    post.highlights = (post.highlights ?? []).map((item) => item?.trim()).filter(Boolean);
  }

  private mapAdmin(post: BlogPost) {
    return {
      id: post.id,
      title: post.title,
      slug: post.slug,
      excerpt: post.excerpt,
      category: post.category,
      coverImageUrl: post.coverImageUrl,
      coverImageObjectName: post.coverImageObjectName ?? null,
      authorName: post.authorName,
      readTimeMinutes: post.readTimeMinutes,
      isFeatured: post.isFeatured,
      isPublished: post.isPublished,
      publishedAt: post.publishedAt ?? null,
      tags: post.tags ?? [],
      highlights: post.highlights ?? [],
      paragraphs: post.paragraphs ?? [],
      createdAt: post.createdAt,
      updatedAt: post.updatedAt,
    };
  }

  private mapSummary(post: BlogPost) {
    return {
      id: post.id,
      title: post.title,
      slug: post.slug,
      excerpt: post.excerpt,
      category: post.category,
      coverImageUrl: post.coverImageUrl,
      authorName: post.authorName,
      publishedAt: post.publishedAt ?? post.updatedAt,
      readTimeMinutes: post.readTimeMinutes,
      isFeatured: post.isFeatured,
    };
  }

  private slugify(title: string) {
    return (title ?? 'blog-post')
      .trim()
      .toLowerCase()
      .replace(/[^a-z0-9\s-]/g, '')
      .replace(/\s+/g, '-')
      .replace(/-+/g, '-')
      .replace(/^-|-$/g, '') || 'blog-post';
  }
}
